test/E2E/Helpers.purs

```diff
@@ -0,0 +1,365 @@
+waitForWalletPage jQuery browser =
+  findWalletPage jQuery browser >>= case _ of
+    Nothing -> do
+      delay $ wrap 100.0
+      waitForWalletPage jQuery browser
+    Just page -> pure page
+
+showOutput :: Ref (Array E2EOutput) -> Effect String
+showOutput ref =
+  let
+    show' (E2EOutput { outputType, output }) = showType outputType <> " " <>
+      output
+    showType PageError = "ERR"
+    showType Console = "..."
+    showType RequestFailed = "REQ"
+  in
+    Ref.read ref >>= pure <<< intercalate "\n" <<< map show'
+
+-- | start example at URL with Nami and return both Nami's page and the example's
+startExample :: String -> Toki.Browser -> Aff RunningExample
+startExample name browser = do
+  page <- Toki.newPage browser
+  jQuery <- retrieveJQuery page
+  errorRef <- liftEffect $ Ref.new []
+  liftEffect do
+    Toki.onPageError (handler errorRef PageError $ pure <<< show) page
+    Toki.onConsole (handler errorRef Console Toki.consoleMessageText) page
+    Toki.onRequestFailed (handler errorRef RequestFailed $ pure <<< show) page
+  Toki.goto url page
+  pure $ wrap
+    { browser: browser
+    , jQuery: jQuery
+    , main: page
+    , errors: errorRef
+    }
+  where
+  url :: Toki.URL
+  url = wrap $ unwrap exampleUrl <> "?" <> name
+
+  handler
+    :: forall (a :: Type)
+     . Ref (Array (E2EOutput))
+    -> OutputType
+    -> (a -> Aff String)
+    -> EffectFn1 a Unit
+  handler errorRef outputType f = mkEffectFn1 $ \e -> launchAff_ do
+    output <- f e
+    liftEffect $ Ref.modify_
+      ( _ <>
+          [ E2EOutput
+              { outputType: outputType
+              , output: output
+              }
+          ]
+      )
+      errorRef
+
+checkSuccess :: RunningExample -> Aff Unit
+checkSuccess (RunningExample { main, errors }) = do
+  feedback <- testFeedbackIsTrue main
+  unless feedback $ liftEffect $ showOutput errors >>= log
+  shouldSatisfy feedback (_ == true)
+
+inWalletPage
+  :: forall (a :: Type). RunningExample -> (Toki.Page -> Aff a) -> Aff a
+inWalletPage (RunningExample { browser, jQuery }) =
+  (waitForWalletPage jQuery browser >>= _)
+
+namiConfirmAccess :: RunningExample -> Aff Unit
+namiConfirmAccess = flip inWalletPage (clickButton "Access")
+
+namiSign :: RunningExample -> Aff Unit
+namiSign = flip inWalletPage \nami -> do
+  clickButton "Sign" nami
+  reactSetValue password testPassword nami
+  clickButton "Confirm" nami
+
+geroConfirmAccess :: RunningExample -> Aff Unit
+geroConfirmAccess =
+  flip inWalletPage $ \page -> do
+    delay $ wrap 100.0
+    void $ doJQ (inputType "radio") click page
+    void $ doJQ (buttonWithText "Continue") click page
+    delay $ wrap 100.0
+    void $ doJQ (inputType "checkbox") click page
+    void $ doJQ (buttonWithText "Connect") click page
+
+geroInit :: Toki.Browser -> Aff Unit
+geroInit browser = do
+  page <- Toki.newPage browser
+  jQuery <- retrieveJQuery page
+  Toki.goto
+    (wrap "chrome-extension://iifeegfcfhlhhnilhfoeihllenamcfgc/index.html")
```

Comment:
I think the URL should be made into a constant or something. It's kind of hidden here and it might not be immediately clear what it's pointing to